File: TidbitsTrivia/Core/Models/Question.swift
import SwiftUI

/// A single trivia question. Produced by the template engine — whether
/// pre-baked into the bundled corpus or generated live from a Wikipedia
/// article. The shape is identical for both paths so the game loop never
/// cares where a question came from.
nonisolated struct Question: Identifiable, Hashable, Codable, Sendable {
    let id: String
    let prompt: String
    let options: [String]      // exactly 4; index 0..3
    let correctIndex: Int
    let categoryID: String
    let difficulty: Int        // 1 (easy) … 5 (hard) — see DifficultyModel
    let explanation: String    // the "learn the fact" payload, shown after answering
    let sourceTitle: String    // Wikipedia article title
    let sourceURL: URL?
    let templateID: String
    var imageURL: URL? = nil   // Picture ID (Q7): the Commons image to identify
    var closest: ClosestSpec? = nil   // Closest Call (M5): numeric estimation
    var ordering: [String]? = nil     // Ordering (Q4): the items in CORRECT order
    var matching: MatchSpec? = nil    // Matching (Q5): keys ↔ correct values
    var accepted: [String]? = nil     // Type-the-answer (Q6): accepted free-text answers

    var correctAnswer: String {
        if options.indices.contains(correctIndex) { return options[correctIndex] }
        return closest?.formattedAnswer ?? ""
    }

    /// Stable share text — never leaks the answer when used pre-reveal.
    func shareTeaser() -> String {
        "🧠 Tidbits Trivia\n\(prompt)\n\nThink you know it? Play at tidbits.trivia"
    }
}

/// Closest Call (M5) numeric question: estimate a value on a linear slider over
/// [min, max]; scored by proximity to `answer` within `tolerance` (adds-only).
nonisolated struct ClosestSpec: Hashable, Codable, Sendable {
    let answer: Double
    let min: Double
    let max: Double
    let step: Double
    let tolerance: Double
    let unit: String

    /// Points (0…maxPoints) for a guess — full at exact, 0 at/over tolerance.
    static let maxPoints = 50
    func points(for guess: Double) -> Int {
        let error = abs(guess - answer)
        guard error < tolerance else { return 0 }
        return Int((Double(Self.maxPoints) * (1 - error / tolerance)).rounded())
    }
    /// "Close enough" to count as correct for streaks / the emoji grid.
    func isClose(_ guess: Double) -> Bool { abs(guess - answer) <= tolerance / 2 }
    var formattedAnswer: String {
        let n = answer == answer.rounded() ? String(Int(answer)) : String(answer)
        return unit.isEmpty ? n : "\(n) \(unit)"
    }
}

/// Matching (Q5): link each key to its value. `values[i]` is the correct match
/// for `keys[i]`; the client shuffles the values for display.
nonisolated struct MatchSpec: Hashable, Codable, Sendable {
    let keys: [String]
    let values: [String]
}

/// Whether the player got it right, and how fast — drives speed scoring,
/// streaks, and the spaced re-ask of missed questions.
nonisolated struct AnsweredQuestion: Identifiable, Hashable, Sendable {
    let question: Question
    let chosenIndex: Int?      // nil = timed out
    let secondsTaken: Double
    var id: String { question.id }
    var isCorrect: Bool { chosenIndex == question.correctIndex }
}

File: windows/Tidbits.Core/Models/Question.cs
using System.Text.Json.Serialization;

namespace Tidbits.Core.Models;

/// A single trivia question — identical shape whether pre-baked in the corpus or
/// generated live from a Wikipedia article (port of Core/Models/Question.swift).
/// JSON keys match the Swift Codable output so the same corpus/wire data loads.
public sealed record Question
{
    [JsonPropertyName("id")] public required string Id { get; init; }
    [JsonPropertyName("prompt")] public required string Prompt { get; init; }
    [JsonPropertyName("options")] public required IReadOnlyList<string> Options { get; init; } // exactly 4; index 0..3
    [JsonPropertyName("correctIndex")] public int CorrectIndex { get; init; }
    [JsonPropertyName("categoryID")] public required string CategoryId { get; init; }
    [JsonPropertyName("difficulty")] public int Difficulty { get; init; } // 1 (easy) … 5 (hard)
    [JsonPropertyName("explanation")] public string Explanation { get; init; } = "";
    [JsonPropertyName("sourceTitle")] public string SourceTitle { get; init; } = "";
    [JsonPropertyName("sourceURL")] public string? SourceUrl { get; init; }
    [JsonPropertyName("templateID")] public string TemplateId { get; init; } = "";
    [JsonPropertyName("tags")] public IReadOnlyList<string> Tags { get; init; } = []; // Wikipedia-category topic keywords
    [JsonPropertyName("imageURL")] public string? ImageUrl { get; init; }        // Picture ID
    [JsonPropertyName("closest")] public ClosestSpec? Closest { get; init; }     // Closest Call
    [JsonPropertyName("ordering")] public IReadOnlyList<string>? Ordering { get; init; } // items in CORRECT order
    [JsonPropertyName("matching")] public MatchSpec? Matching { get; init; }     // Matching
    [JsonPropertyName("accepted")] public IReadOnlyList<string>? Accepted { get; init; } // Type-the-answer
    [JsonPropertyName("enumerate")] public EnumSpec? Enumerate { get; init; }    // Enumeration
    [JsonPropertyName("roundIndex")] public int? RoundIndex { get; init; }       // Trivia Night round (runtime)

    [JsonIgnore]
    public string CorrectAnswer =>
        CorrectIndex >= 0 && CorrectIndex < Options.Count
            ? Options[CorrectIndex]
            : Closest?.FormattedAnswer ?? "";
}

/// Closest Call numeric question: estimate a value on a linear slider over
/// [min, max]; scored by proximity to `answer` within `tolerance` (adds-only).
public sealed record ClosestSpec
{
    [JsonPropertyName("answer")] public double Answer { get; init; }
    [JsonPropertyName("min")] public double Min { get; init; }
    [JsonPropertyName("max")] public double Max { get; init; }
    [JsonPropertyName("step")] public double Step { get; init; }
    [JsonPropertyName("tolerance")] public double Tolerance { get; init; }
    [JsonPropertyName("unit")] public string Unit { get; init; } = "";

    public const int MaxPoints = 50;

    /// Points (0…MaxPoints) — full at exact, 0 at/over tolerance.
    public int Points(double guess)
    {
        var error = Math.Abs(guess - Answer);
        if (error >= Tolerance) return 0;
        return (int)Math.Round(MaxPoints * (1 - error / Tolerance));
    }

    /// "Close enough" to count as correct for streaks / the emoji grid.
    public bool IsClose(double guess) => Math.Abs(guess - Answer) <= Tolerance / 2;

    [JsonIgnore]
    public string FormattedAnswer
    {
        get
        {
            var n = Answer == Math.Round(Answer)
                ? ((long)Answer).ToString()
                : Answer.ToString(System.Globalization.CultureInfo.InvariantCulture);
            return string.IsNullOrEmpty(Unit) ? n : $"{n} {Unit}";
        }
    }
}

/// Matching: values[i] is the correct match for keys[i]; the client shuffles for display.
public sealed record MatchSpec
{
    [JsonPropertyName("keys")] public required IReadOnlyList<string> Keys { get; init; }
    [JsonPropertyName("values")] public required IReadOnlyList<string> Values { get; init; }
}

/// Enumeration: each group is one accepted answer ([canonical, alias, …]); a group
/// fills only once; the display name is group[0].
public sealed record EnumSpec
{
    [JsonPropertyName("groups")] public required IReadOnlyList<IReadOnlyList<string>> Groups { get; init; }
    [JsonIgnore] public int Total => Groups.Count;
    [JsonIgnore] public IReadOnlyList<string> DisplayNames => Groups.Select(g => g.Count > 0 ? g[0] : "").ToList();
}

/// Whether the player got it right, and how fast — drives speed scoring, streaks,
/// and spaced re-ask of missed questions.
public sealed record AnsweredQuestion
{
    public required Question Question { get; init; }
    public int? ChosenIndex { get; init; } // null = timed out
    public double SecondsTaken { get; init; }
    public bool IsCorrect => ChosenIndex == Question.CorrectIndex;
}
